Documentation for using OTP with an existing fastapi app. (#666)
Materials-Consortia/optimade-python-tools: docs/use_cases/integrated.md

# Integrate with an existing web application

> OPTIMADE Python tools (OPT) can be run as a standalone web application that serves the OPTIMADE API based on a pre-configured MongoDB backend.

Reviewer: ```suggestion
`optimade-python-tools` (OPT) can be run as a standalone web application that serves the OPTIMADE API based on a pre-configured MongoDB backend.
```

> In this document, we are going to use OPT differently and use it to add an OPTIMADE API implementation alongside an existing API employing an Elasticsearch storage layer.

Reviewer: ```suggestion

This is the approach taken by the [NOMAD repository](http://nomad-lab.eu/prod/rae/gui/) to host their [OPTIMADE API](https://nomad-lab.eu/prod/rae/optimade/).
This library is used to filter millions of structures from published first-principles calculations provided by users and other providers.
NOMAD also uses the filtering module in its own API to expose the OPTIMADE filter language in the user-centric web interface search bar.

In this document, we are going to use OPT differently and use it to add an OPTIMADE API implementation alongside an existing API employing an Elasticsearch storage layer.
```

> In this document, we are going to use OPT differently and use it to add an OPTIMADE API implementation to an existing application with an existing Elasticsearch-based database.

Reviewer: ```suggestion
In this document, we are going to use OPT differently and use it to add an OPTIMADE API implementation alongside an existing API employing an Elasticsearch storage layer.
```

> Let's assume we already have a *FastAPI* application that runs an unrelated web service, and that we use an Elasticsearch backend that contains all structure data, but not necessarily into a form that OPT expects.

Reviewer: ```suggestion
Let's assume we already have a *FastAPI* application that runs an unrelated web service, and that we use an Elasticsearch backend that contains all structure data, but not necessarily into a form that OPTIMADE expects.
```

> ## Providing an OPTIMADE Python tools configuration

Reviewer: ```suggestion
## Providing an `optimade-python-tools` configuration
```

> OPT reads its configuration from a file.

Reviewer: ```suggestion
OPT can read its configuration from a JSON file.
```

os.environ['OPTIMADE_CONFIG_FILE'] = os.path.join(
    os.path.dirname(__file__), 'optimade_config.json'
)
```

## Customize the [`EntryCollection`][optimade.server.entry_collections.entry_collections.EntryCollection] implementation

Let's assume that your Elasticsearch backend stores structure data different enough that you need to provide your own custom implementation.
The following code customizes the used [`EntryCollection`][optimade.server.entry_collections.entry_collections.EntryCollection] class for structures, while keeping the default *mongomock* implementation (using [`MongoCollection`][optimade.server.entry_collections.mongo.MongoCollection]) for all other entry types.

```python
from optimade.server.routers import structures

structures.structures_coll = MyElasticsearchStructureCollection()
```

You can imagine that `MyElasticsearchStructureCollection` either sub-classes the default OPT Elasticsearch implementation or sub-classes `EntryCollection`, depending on how deeply you need to customize the default OPT behavior.

## Mounting the OPTIMADE Python tools *FastAPI* app into an existing *FastAPI* app

Let's assume you have an existing *FastAPI* app `my_app`.

> It already adds a few routers under certain path prefixes and now you want to add the OPT API implementation under the path prefix `/optimade`.

Reviewer: ```suggestion
It already adds a few routers under certain path prefixes and now you want to add an OPTIMADE implementation under the path prefix `/optimade`.
```

# Integrate with an existing web application

`optimade-python-tools` (OPT) can be run as a standalone web application that serves the OPTIMADE API based on a pre-configured MongoDB backend.
In this document, we are going to use OPT differently and use it to add an OPTIMADE API implementation alongside an existing API employing an Elasticsearch storage layer.

> Let's assume we already have a *FastAPI* application that runs an unrelated web service, and that we use an Elasticsearch backend that contains all structure data, but not necessarily into a form that OPTIMADE expects.

Reviewer 2: ```suggestion
Let's assume we already have a *FastAPI* application that runs an unrelated web service, and that we use an Elasticsearch backend that contains all structure data, but not necessarily in a form that OPTIMADE expects.
```
I think "in" is the correct preposition here.
